Single-Select Button Entry in Web Apps

nospame opened the pull request:
## Product Description
Using appearance attribute `button-select`, single-select multiple choice questions in web apps can be formatted as simple buttons. Most likely use case is for a display condition to show/hide questions on button click.

https://github.com/dimagi/commcare-hq/assets/100609770/3ddfd776-c52a-4145-aca1-447e36a2b78f

When using alignment appearance attributes `text-align-center` and `text-align-right`, aligns based on the available width or other entry fields, not labels, for better custom positioning.
Larger screen:
![image](https://github.com/dimagi/commcare-hq/assets/100609770/e2cf7606-d664-4a3c-96b7-4c08996efae0)

Smaller screen:
![image](https://github.com/dimagi/commcare-hq/assets/100609770/58ae2198-4c22-42c3-9c85-2637089cabdb)


## Technical Summary
[USH-3827](https://dimagi-dev.atlassian.net/browse/USH-3827)
Adds a `ButtonSelectEntry` which uses a simplified template lacking any label or validation info, to show a button that cycles through the available "select" choices. The text on the button is the next selection, indicating what will happen when a user clicks the button.


## Safety Assurance

### Safety story
Tested locally with display conditions dependent on the button value, as well as formatting with other appearance attributes `n-per-row`, `text-align-center` and `text-align-right`.

### Automated test coverage
Added JS tests for `ButtonSelectEntry` and option-cycling logic.

### QA Plan
This probably wants QA to go through circumstances I haven't considered, test with other appearance attributes, etc.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-3827]: https://dimagi-dev.atlassian.net/browse/USH-3827?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Create new ButtonSelectEntry w button-select appearance attribute

nospame committed:
Don't show label text or validation for button select questions

nospame committed:
Tests for ButtonSelectEntry

nospame commented on corehq/apps/cloudcare/templates/form_entry/templates.html:
```diff
@@ -201,7 +201,8 @@ <h1 class="title" data-bind="text: title, visible: !showInFormNavigation()"></h1
   <div data-bind="class:questionTileWidth">
-  <!-- ko if: datatype() !== 'info' -->
+  <!-- ko if: datatype() !== 'info'
+    && !(datatype() === 'select' && stylesContains('button-select')) -->
```
This conditional is really ugly. I considered adding it [here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js#L730) as a property on the question, but after seeing that done with `self.isLabel` and that _not_ being used as the knockout conditional here, I sort of assumed there was a good reason (avoiding any semblance of circular dependencies, maybe?)

Jtang-1 reviewed (approved):
nothing blocking. Looks like a clean implementation

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:
```diff
@@ -523,6 +523,32 @@ hqDefine("cloudcare/js/form_entry/entries", function () {
+     * Represents a single button that cycles through choices
+     */
+    function ButtonSelectEntry(question, options) {
+        var self = this;
```
It could even be `const` since it's not getting redefined, but I actually prefer `var` here because that's how other entry classes are set up. There's not a _ton_ of consistency of variable types in that file, but I also think it would be out of scope here to redefine them all in this PR given it's such a key piece of web apps.

nospame committed:
Refactor: simplify entry styles checks

Add isButton to form_ui for template check. Remove unused style var in getEntry. Clean up conditional checks for select and multi-select entries

nospame committed:
Merge branch 'jls/web-apps-label-style' of github.com:dimagi/commcare-hq into em/button-select-entry+jls/web-apps-label-style

Jtang-1 commented on corehq/apps/cloudcare/templates/form_entry/templates.html:
```diff
@@ -285,7 +284,7 @@ <h1 class="title" data-bind="text: title, visible: !showInFormNavigation()"></h1
                     }">{% trans 'Sorry, this response is required!' %}</div>
   <!-- /ko -->
-  <!-- ko if: datatype() === 'info' -->
+  <!-- ko if: isLabel -->
```
This comment is nonblocking and not a result of this PR so I don't expect any changes but I think it's confusing how I think this instance of "isLabel" is checking if the entry is an instance of `InfoEntry` [here](https://github.com/dimagi/commcare-hq/blob/1cefe1ec6d1f3fc8dd65cd4af9bd35b5c7f90cff/corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js#L1319). But in `getEntry`, there's other checks that's named `isLabel` which checks the style, and not related to `datatype()==='info'`. I found that confusing, and i think it would be nice to clear up. 

This instance is fine since it's pulling from  the [Question] attribute (https://github.com/dimagi/commcare-hq/blob/946e797a1dc88d44cff2112da5189fdeae9b2004/corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js#L768)

nospame replied:
Good callout, thank you. I'll look into the two different definitions and see if it makes sense to rename one of them here for clarity.

Jtang-1 replied:
⭐

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/button-select-entry

nospame committed:
Fix button select alignment - align with fields and not labels

nospame committed:
Render markdown on button select entry

nospame committed:
Refactor: isLabel => isChoiceLabel where used to render a ChoiceLabelEntry, remove unused is_select property

Jtang-1 reviewed (approved):
Just for accuracy, could you update the video in the description with the new alignment?
